// route42/Router.js
var Mod1 = require("mod42/Mod1");
var Mod2 = require("mod42/Mod2");
var Mod3 = require("mod42/Mod3");
var is = require("is");
var Route = require("./Route");
var History = require("history").createBrowserHistory;

var logger = require("log42");
var log = logger();

var RouterView = require("./RouterView");

var Router = module.exports = Mod3.extend({
	log: true,
	name: "Router",
	Route: Route,
	View: RouterView,
	init: function(){
		this.cbs = [];
		this.dcbs = [];
		this.routes = [];
		this.history = History();
		this.history.listen(this.historyListener.bind(this));
	},
	historyListener: function(location, action){
		var match;
		if (action === "POP"){
			match = this.match();
			if (match){
				this.log("match found");
				match.activate();
			}
			else
				this.log("no match")
		} else {
			this.log("action:", action, "(do nothing)");
		}
	},
	matchAndActivate: function(){
		var match = this.match();
		if (match)
			match.activate();
		return this;
	},
	match: function(){
		var pathParts = this.getPathParts(), match;
		for (var i = 0; i < this.routes.length; i++){
			match = this.routes[i].match(pathParts)
			if (match)
				return match;
		}
		return false;
	},
	getPathParts: function(){
		return this.cleanPathParts(window.location.pathname.split("/"));
	},
	cleanPathParts: function(pathParts){
		if (pathParts[0] === "")
			pathParts = pathParts.slice(1);

		if (pathParts[pathParts.length - 1] === "")
			pathParts = pathParts.slice(0, -1);

		return pathParts;
	},
	// just for root routes?
	add: function(route){
		if (!(route instanceof Route)){
			route = Route(route);
		}

		route.router = this;
		this.routes.push(route);

		if (!this[route.name]){
			this[route.name] = route;
		}

		return route;
	}
});
