// Application/Application.js
var utils = require("utils");
var is = utils.is;

var global_app_module = require("app");
var registered_global_module = false;

var Page = require("Page/Page2");
var Router = require("core/Router");
var View = require("View");

var Mod1 = require("mod42/Mod1");
var Mod2 = require("mod42/Mod2");

var $ = require("jquery");

var App = Mod2.extend({
	name: "App",
	Router: Router,
	Page: Page,
	View: View,
	init: function(){
		this.init_app();
	},
	init_app: function(){
		this.init_modules();
		this.register_global_module();
		this.init_router();
		this.load_pages();
		this.load_tests();
	},
	init_modules: function(){
		this.Router = this.Router.extend({
			name: "AppRouter",
			log: this.log.value,
			app: this
		});

		this.Page = this.Page.extend({
			name: "AppPage",
			log: this.log.value,
			app: this
		});
	},
	init_router: function(){
		this.router = this.Router();
		
		this.router.render_nav();
	},
	register_global_module: function(){
		if (!registered_global_module){
			global_app_module.exports = this;
			registered_global_module = true;
		}
	},
	load_pages: function(){
		// this has to be done from within the site?
		var pages = require.context("../", true, /\.page\.js$/);
		this.require_all(pages);
	},
	require_all: function(req_ctx){
		req_ctx.keys().forEach(req_ctx);
	}
});

console.assert(App.prototype.init.wrapped);

module.exports = App;
