// View/Page.js
var View = require("./View");

var Page = View.Page = module.exports = View.extend({
	name: "Page",
	autoRender: false,
	addClass: "page",
	css: {
		marginBottom: "1em",
		background: "rgba(0,0,0,0.1)"
	},
	size: 0,
	depth: 1,
	inst: function(){
		this.pages = [];
	},
	content: function(){
		var page = this;

		View({
			tag: "button",
			content: "back",
			click: [function(){
				page.pager.back()
			}]
		})

		View({
			tag: "h3",
			content: this.name
		});

		View(function(){
			this.filler("1-4p");
		});

		this.generate();
	},
	generate: function(){
		var page = this;
		if (this.size && this.depth > 0){
			View({
				addClass: "children",
				css: {
					padding: "10px"
				},
				content: function(){
					for (var i = 1; i <= page.size; i++){
						new page.SubPage({
							name: page.name + "-" + i,
							parent: page,
							pager: page.pager,
							size: page.size,
							depth: page.depth - 1
						}).hide().render().button()
					}
				}
			})
		}
	},
	button: function(){
		var page = this;
		View(function(){
			this.append(page.name);
			this.css({
				padding: "6px 12px",
				cursor: "pointer"
			});
			this.click(function(){
				// console.log("click");
				page.activate();
			})
		})
	},
	activate: function(){
		this.pager.to(this);
	}
});

Page.prototype.SubPage = Page.extend({
	name: "SubPage"
});
