// mod42/Mod2.js
var SetMfn = require("./SetMfn");
var logger = require("log42");
var is = require("is");
var Mod1 = require("./Mod1");

var Method = require("log42/Method");
var Property = require("Property");

var log = logger();

log.groupc("Mod2.js");

log.group("var set = SetMfn.make(...)");
var set = SetMfn.make({
	stdProp: function(mod, prop, value){
		if (is.fn(value) && !is.Class(value))
			this.wrapMethod(mod, prop, value);
		else
			// this.wrapProperty(mod, prop, value);
			mod[prop] = value;
	},
	Method: Method,
	wrapMethod: function(mod, name, fn, opts2){
		// console.log("wrappingMethod", name);
		var opts = this["method_" + name];
		if (opts === false){
			mod[name] = fn.fn || fn;
		} else {
			opts = opts || {};
			mod[name] = new this.Method({
				name: name,
				method: fn.fn || fn,
			}, opts, opts2).wrapper();
		}

	},
	Property: Property,
	wrapProperty: function(mod, name, value){
		new this.Property({
			name: name,
			parent: mod,
			value: value
		});
	},
	method__config: false
});
log.end();


log.group("var Mod2 = Mod1.extend(...)");
var Mod2 = module.exports = Mod1.extend({
	name: "Mod2",
	assign: {
		set: set
	},
	_config: function(o){
		if (o && o.log){
			this.log = o.log;
			delete o.log;
		}

		if (o && o.name){
			this.name = o.name;
			delete o.name;
		}
	}
}).assign({
	set: set
});

// console.log(Mod2.prototype.set.mfn.stdProp.toString());

logger.install(Mod2.prototype);
logger.install(Mod2);

set.mfn.wrapMethod(Mod2, "extend", Mod2.extend);
set.mfn.wrapMethod(Mod2, "set", Mod2.set);

set.mfn.wrapMethod(Mod2.prototype, "instantiate", Mod2.prototype.instantiate);
set.mfn.wrapMethod(Mod2.prototype, "initialize", Mod2.prototype.initialize);
set.mfn.wrapMethod(Mod2.prototype, "init", Mod2.prototype.init);
set.mfn.wrapMethod(Mod2.prototype, "set", Mod2.prototype.set);

log.end();
log.end();
