// mod42/SetMfn2.js
var Mfn = require("mod42/Mfn");
var ExtendMfn = require("mod42/ExtendMfn");
var is = require("is");
var logger = require("log42");

var log = logger();

log.groupc("SetMfn2.js");

var SetMfn2 = module.exports = Mfn.extend({
	name: "SetMfn2",
	instantiate: function(){
		if (this.set)
			this.set.apply(this, arguments);
		else
			this.assign.apply(this, arguments);
		this.init();
	},
	main: function(mod){
		for (var i = 1; i < arguments.length; i++){
			this.arg(mod, arguments[i]); // keep mod as first arg
		}
		this.then(mod);
		return mod; // important
	},
	arg: function(mod, arg, parent){
		if (is.pojo(arg))
			this.obj(mod, arg, parent);
		else
			this.other(mod, arg, parent);
	},
	std: function(current, value, parent, name){
		// parent[name] === current
		// current + value ==> ???
		this.log.group(prop, this.log.inline(value));

		// coordinate this with the mod.props...
		if (this.props[prop]){
			this.props[prop].apply(this, arguments);

		} else if (is.undef(current)){
			this.stdProp.apply(this, arguments);

		} else if (is.Class(current)) {
			this.class.apply(this, arguments);

		} else if (currentValue.clone){
			this.stdProp(mod, prop, currentValue.clone(value));

		} else if (currentValue.set) { // dependent on is.def(currentValue)
			this.propWithSet(mod, prop, value);

		} else if (is.fn(currentValue)) {
			this.fnProp(mod, prop, value);

		} else if (is.obj(currentValue)){
			this.objProp(mod, prop, value);

		} else {
			this.stdProp(mod, prop, value);
		}

		this.log.end();
	},
	assignment: function(parent, name, value, current){
		parent[name] = value;
	},
	class: function(current, value, parent, name){

		this.assignment(parent, name, value, current);

		if (is.Class(value)){
			this.stdProp(mod, prop, value);
		} else if (mod.hasOwnProperty(prop)) {
			mod[prop].prototype.set(value);
		} else {
			mod[prop] = mod[prop].extend(value); // protect mod's prototype
		}
	},
	other: function(mod, arg, parent){
		if (is.def(arg))
			console.warn("not sure how to set this", arg);
	},
	obj: function(mod, obj, parent){
		for (var i in obj){
			this.prop(mod, i, obj[i]);
		}
		return mod; // used in .objProp()
	},
	props: {
		log: function(mod, prop, value){
			logger.install(mod, value);
		},
		parent: function(mod, prop, value){
			this.stdProp(mod, prop, value);
		}
	},
	// this is like (parent, name, arg)
	// and "mod", is like the current value
	// then there's the parent[name] === current value
	// and then then currentValue[name] => new value / setters
	prop: function(mod, prop, value){
		this.log.group(prop, this.log.inline(value));

		var currentValue = mod[prop];

		// coordinate this with the mod.props...
		if (this.props[prop]){
			this.props[prop].call(this, mod, prop, value);

		} else if (is.undef(currentValue)){
			this.stdProp(mod, prop, value);

		} else if (is.Class(currentValue)) {
			this.ClassProp(mod, prop, value);

		} else if (currentValue.clone){
			this.stdProp(mod, prop, currentValue.clone(value));

		} else if (currentValue.set) { // dependent on is.def(currentValue)
			this.propWithSet(mod, prop, value);

		} else if (is.fn(currentValue)) {
			this.fnProp(mod, prop, value);

		} else if (is.obj(currentValue)){
			this.objProp(mod, prop, value);

		} else {
			this.stdProp(mod, prop, value);
		}

		this.log.end();
	},
	propWithSet: function(mod, prop, value){
		if (!mod.hasOwnProperty(prop) && !mod[prop].mfn){
			this.stdProp(mod, prop, Object.create(mod[prop]).set(value));
		} else {
			this.stdProp(mod, prop, mod[prop].set(value));
		}
	},
	ClassProp: function(mod, prop, value){
		if (is.Class(value)){
			this.stdProp(mod, prop, value);
		} else if (mod.hasOwnProperty(prop)) {
			mod[prop].prototype.set(value);
		} else {
			mod[prop] = mod[prop].extend(value); // protect mod's prototype
		}
	},
	stdProp: function(mod, prop, value){
		mod[prop] = value;

		if (value && is.fn(value.setup))
			value.setup(mod, prop);
	},
	fnProp: function(mod, prop, value){
		if (is.fn(value)){
			this.stdProp(mod, prop, value);
		} else {
			if (is.arr(value))
				mod[prop].apply(mod, value);
			else
				mod[prop].call(mod, value);
		}
	},
	objProp: function(mod, prop, value){
		if (is.obj(value)){
			if (mod.hasOwnProperty(prop))
				this.obj(mod[prop], value)
			else
				this.stdProp(mod, prop, this.obj(Object.create(mod[prop]), value));
		} else {
			console.warn("whoops");
		}
	},
	then: function(){}
}).assign({
	extend: ExtendMfn.make()
});

var set = SetMfn2.make();

SetMfn2.set = set;
SetMfn2.prototype.set = set;

log.end();
