// Page/Page2.js
var View = require("View");

var is = require("is");
var Route = require("core/Route");
var utils = require("utils");
var Base3 = require("Base3");
var Link = require("./Link");

var $ = require("jquery");

var Page2 = View.extend({
	Route: Route,
	Link: Link,
	name: "Page2",
	addClass: "page",
	autoRender: false,
	init: function(){
		this.init_page();

		// auto render
		this.init_view();
	},
	init_page: function(){
		this.app[this.name] = this;
		this.init_route();
	},
	init_route: function(){
		var page = this;
		this.route = this.Route({
			pathname: this.pathname || utils.sanitizeString(this.name),
			label: this.title || this.name
		}).then(function(){
			// console.log("page.then");
			$(function(){
				page.render();
				page.$el.appendTo("#docroot");
				// console.log("page.then --> ready");
			});
		}).andThen(function(){
			page.remove();
		});

		this.app.router.addRoutes(this.route);
	},
	link: function(){
		return this.Link({
			page: this
		});
	}
});

module.exports = Page2;
