// app42/App.js
var logger = require("log42");
var log = logger();

log.groupc("App.js");

var utils = require("utils");
var is = utils.is;

var global_app_module = require("app");
var registered_global_module = false;

var View = require("View");
var Page = require("Page/Page2");
var router = require("route42");

var Mod1 = require("mod42/Mod1");
var Mod2 = require("mod42/Mod2");
var Mod3 = require("mod42/Mod3");

var $ = require("jquery");

var App = module.exports = Mod3.extend({
	methods: {
		instantiate: {
			expand: true
		}
	},
	// log: true,
	name: "App",
	Page: Page,
	View: View,
	initialize: function(){
		this.init_app();
		this.init();
	},
	init_app: function(){
		// this.register_global_module();
		this.init_modules();
		this.init_router();
		this.load_pages();
		this.load_tests();
	},
	init_modules: function(){
		this.Page = this.Page.extend({
			name: "AppPage",
			log: this.log.value,
			app: this
		});
	},
	init_router: function(){
		this.router = router;
	},
	register_global_module: function(){
		if (!registered_global_module){
			global_app_module.exports = this;
			registered_global_module = true;
		}
	},
	load_pages: function(){},
	load_tests: function(){},
	require_all: function(req_ctx){
		req_ctx.keys().forEach(function(v){
			this.log.groupc("load", v);
			req_ctx(v);
			this.log.end();
		}.bind(this));
	}
});

log.end();
