// Property/Property.js
var Base3 = require("Base3");
var is = require("is");

var Property = Base3.extend({
	name: "Property",
	value: undefined,
	init: function(){
		this.init_define();
		this.init_attach();
	},
	init_define: function(){
		var name = this.name;

		Object.defineProperty(this.parent, this.name, {
			get: function(){
				return this.props[name].getter();
			},
			set: function(value){
				if (this.hasOwnProperty(name)){
					this.props[name].setter(value);
				} else {
					new Property({
						name: name,
						parent: this,
						value: value
					});
				}
			},
			configurable: true
		});
	},
	init_attach: function(){
		this.parent.props = this.parent.props || {};
		this.parent.props[this.name] = this;
	},
	getter: function(){
		console.log(this.parent.name + "." + this.name + " -->", this.value);
		return this.value;
	},
	setter: function(value){
		console.log(this.parent.name + "." + this.name + " =", value);
		this.value = value;
	}
});

// if (this.constructor.prototype === this).. use ProtoProp, otherwise use Property...?

module.exports = Property;
